File: RepoTransBench/source_projects/Python/pallets_itsdangerous/tests/test_itsdangerous/test_timed.py
import time
import pytest
from itsdangerous import TimedSerializer, SignatureExpired, BadTimeSignature, BadSignature

def test_timedserializer_dumps_loads():
    s = TimedSerializer("secret-key")
    data = {"msg": "timed"}
    dumped = s.dumps(data)
    loaded = s.loads(dumped)
    assert loaded == data

def test_timedserializer_expiry():
    s = TimedSerializer("secret-expiry")
    data = {"foo": 1}
    token = s.dumps(data)
    assert s.loads(token) == data
    # Should be expired, using max_age=0 and a past timestamp
    with pytest.raises(SignatureExpired):
        s.loads(token, max_age=-1)

def test_timedserializer_tuple_loading_options():
    s = TimedSerializer("secret-key2")
    token = s.dumps({"a": 2})
    # loads(..., return_timestamp=True) returns (data, timestamp)
    result = s.loads(token, return_timestamp=True)
    assert isinstance(result, tuple)
    assert result[0] == {"a": 2}
    # Accepting both float and datetime for timestamp, depending on itsdangerous version
    import datetime
    assert isinstance(result[1], (float, datetime.datetime))

def test_timedserializer_bad_signature():
    s = TimedSerializer("secret")
    # A completely broken token, missing '.' separator
    with pytest.raises(BadSignature):
        s.loads("bad-token")

    # Validly signed but purposely altered token
    orig = s.dumps({"foo": 42})
    tampered = orig[::-1]  # Definitely not the same!
    with pytest.raises(BadTimeSignature):
        s.loads(tampered)

File: RepoTransBench/target_projects/Python/Rust/pallets_itsdangerous/tests/original/test_timed.rs
use itsdangerous_rs::timed::*;
use std::collections::HashMap;
use std::time::{Duration, Instant};

#[test]
fn test_timedserializer_dumps_loads() {
    let s = TimedSerializer::new("secret-key");
    let mut data = HashMap::new();
    data.insert("msg", "timed");
    let dumped = s.dumps(&data);
    let (loaded, _): (HashMap<String, String>, _) = s.loads(&dumped, None, false).unwrap();
    assert_eq!(loaded, data);
}

#[test]
fn test_timedserializer_expiry() {
    let s = TimedSerializer::new("secret-expiry");
    let mut data = HashMap::new();
    data.insert("foo", 1);
    let token = s.dumps(&data);
    let (loaded, _): (HashMap<String, i32>, _) = s.loads(&token, None, false).unwrap();
    assert_eq!(loaded, data);

    let expired = s.loads::<HashMap<String, i32>>(&token, Some(-1), false);
    assert!(expired.is_err());
}

#[test]
fn test_timedserializer_tuple_loading_options() {
    let s = TimedSerializer::new("secret-key2");
    let mut map = HashMap::new();
    map.insert("a", 2);
    let token = s.dumps(&map);
    let (result, timestamp) = s.loads::<HashMap<String, i32>>(&token, None, true).unwrap();
    assert_eq!(result, map);
    assert!(timestamp.is_some());
}

#[test]
fn test_timedserializer_bad_signature() {
    let s = TimedSerializer::new("secret");
    // Bad token
    let broken = "bad-token";
    let res = s.loads::<HashMap<String, i32>>(broken, None, false);
    assert!(res.is_err());
    // Tampered
    let mut orig_map = HashMap::new();
    orig_map.insert("foo", 42);
    let orig = s.dumps(&orig_map);
    let tampered: String = orig.chars().rev().collect();
    let res2 = s.loads::<HashMap<String, i32>>(&tampered, None, false);
    assert!(res2.is_err());
}